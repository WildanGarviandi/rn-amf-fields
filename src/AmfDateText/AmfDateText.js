import React, { Component } from 'react'
import PropTypes from 'prop-types'
import { StyleSheet, Text, View } from 'react-native'
import AmfDropdown from '../AmfDropdown/AmfDropdown'
import moment from 'moment'

class AmfDateText extends Component {
  constructor(props) {
    super(props)

    const initialState = {
      valid: false,
      errorMessage: ''
    }

    if (this.props.value != '' && this.props.value != null && this.props.value != undefined) {
      console.log(this.props.value)
      let temp = this.props.value.split('-')
      this.state = {
        ...initialState,
        day: temp[2],
        month: temp[1],
        year: temp[0]
      }
    } else {
      this.state = {
        ...initialState,
        day: '',
        month: '',
        year: ''
      }
    }

    this.dayOptions = this.createDayOptions()
    this.monthOptions = this.createMonthOptions()
    this.yearOptions = this.createYearOptions()
  }

  componentDidMount() {
    this.validate()
  }

  componentDidUpdate(prevProps) {
    if (this.props.value !== prevProps.value) {
      this.validate()
    }
  }

  createDayOptions = () => {
    let opt = []
    for (let i = 1; i <= 31; i++) {
      let str = i.toString()
      let strPadded = str.length === 1 ? '0' + str : str
      opt.push({ label: strPadded, value: strPadded })
    }

    return opt
  }

  createMonthOptions = () => {
    let opt = []
    for (let i = 1; i <= 12; i++) {
      let str = i.toString()
      let strPadded = str.length === 1 ? '0' + str : str
      opt.push({ label: monthNames[i-1], value: strPadded })
    }

    return opt
  }

  createYearOptions = () => {
    let opt = []
    for (let i = 1940; i <= 2017; i++) {
      let str = i.toString()
      opt.push({ label: str, value: str })
    }

    return opt
  }

  updateDate(type, val) {
    let date
    let year = type == 'year' ? val : this.state.year
    let month = type == 'month' ? val : this.state.month
    let day = type == 'day' ? val : this.state.day

    this.setState({ day, month, year })

    if (day == '' || month == '' || year == '') return 'incomplete'

    do {
      date = moment(year + '-' + month + '-' + day, 'YYYY-MM-DD').format('YYYY-MM-DD')
      if (date == 'Invalid date') {
        let dayInt = parseInt(day)
        day = (dayInt-1).toString().padStart(2, '0')
      }
    } while (date == 'Invalid date')

    return date
  }

  inputValid = () => {
    const { validation, label } = this.props
    if (!validation) return ({status: true})

    // check if pristine
    if (this.props.value === null || this.props.value === undefined) return ({status: true})

    // check whether input is required
    if (validation.required) {
      if (this.props.value === '') return ({status: false, message: label + ' tidak boleh kosong'})
    }

    return ({status: true})
  }

  validate = () => {
    let validationObject = this.inputValid()
    this.setState({valid: validationObject.status, errorMessage: validationObject.message})
  }

  render() {
    let dayObject = this.dayOptions.find((item) => item.value == this.state.day)
    let monthObject = this.monthOptions.find((item) => item.value == this.state.month)
    let yearObject = this.yearOptions.find((item) => item.value == this.state.year)

    let validationStyle = {}
    let validationText
    if (!this.state.valid) {
      validationStyle = {borderColor: 'red', borderWidth: 1}
      validationText = <Text style={style.errorText}>{this.state.errorMessage}</Text>
    }

    return (
      <View>
        <View style={ this.props.layout === 'horizontal' ? style.container : null}>
          <Text style={style.label}>{this.props.label}</Text>
          <View style={[style.pickerContainer, validationStyle]}>
            <AmfDropdown
              layout='horizontal'
              item={this.dayOptions}
              value={dayObject ? dayObject.value : null}
              onSelect={(val) => {
                let temp = val
                date = this.updateDate('day', temp)
                if (date != 'incomplete') this.props.onChange(date)
              }}
              selector={
                <View style={[style.selector, {width: 50}]}>
                  <Text style={style.textInput}>{dayObject ? dayObject.label : 'DD'}</Text>
                </View>
              }
            />
            <Text style={style.dividerText}>/</Text>
            <AmfDropdown
              layout='horizontal'
              item={this.monthOptions}
              value={monthObject ? monthObject.value : null}
              onSelect={(val) => {
                let temp = val
                date = this.updateDate('month', temp)
                if (date != 'incomplete') this.props.onChange(date)
              }}
              selector={
                <View style={[style.selector, {width: 50}]}>
                  <Text style={style.textInput}>{monthObject ? monthObject.label : 'MM'}</Text>
                </View>
              }
            />
            <Text style={style.dividerText}>/</Text>
            <AmfDropdown
              layout='horizontal'
              item={this.yearOptions}
              value={yearObject ? yearObject.value : null}
              onSelect={(val) => {
                let temp = val
                date = this.updateDate('year', temp)
                if (date != 'incomplete') this.props.onChange(date)
              }}
              selector={
                <View style={[style.selector, {width: 70}]}>
                  <Text style={style.textInput}>{yearObject ? yearObject.label : 'YYYY'}</Text>
                </View>
              }
            />
          </View>
        </View>
        { validationText }
      </View>
    )
  }
}

const monthNames = ['Jan', 'Feb', 'Mar', 'Apr', 'Mei', 'Jun', 'Jul', 'Ags', 'Sep', 'Okt', 'Nov', 'Des']

const style = StyleSheet.create({
  container : {
    flexDirection: 'row',
    alignItems: 'center',
    // paddingVertical: 8
  },
  dividerText: {
    fontSize: 20,
    color: '#9B9B9B',
    marginHorizontal: 4
  },
  errorText: {
    color: 'red',
    margin: 4,
    marginBottom: 0
  },
  label: {
    color: '#454545',
    marginRight: 16
  },
  pickerContainer: {
    flexDirection: 'row',
    alignItems: 'center',
    padding: 8
  },
  selector: {
    flexDirection: 'row',
    justifyContent: 'center',
    alignItems: 'center',
    height: 30
  },
  textInput: {
    textAlign: 'center'
  }
})

AmfDateText.defaultProps = {
  layout: 'vertical'
}

AmfDateText.propTypes = {
  label: PropTypes.string,
  layout: PropTypes.oneOf(['vertical', 'horizontal']),
  onChange: PropTypes.func.isRequired,
  value: PropTypes.string
}

export default AmfDateText
