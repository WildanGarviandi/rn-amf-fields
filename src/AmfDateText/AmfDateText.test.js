import React from 'react'
import AmfDateText from './AmfDateText'
import renderer from 'react-test-renderer'

test('AmfDateText without validation', () => {
  let props = {
    label: 'Label',
    onChange: () => {},
    value: null
  }

  const component = renderer.create(<AmfDateText {...props} />)

  let tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})

