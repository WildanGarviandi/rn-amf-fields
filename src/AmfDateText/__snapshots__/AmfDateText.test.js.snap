// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AmfDateText without validation 1`] = `
<View>
  <View
    style={null}
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      disabled={false}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#454545",
          "marginRight": 16,
        }
      }
    >
      Label
    </Text>
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "padding": 8,
          },
          Object {},
        ]
      }
    >
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <View
            animated={true}
            defaultIndex={-1}
            defaultValue="Please select..."
            disabled={false}
            dropdownStyle={
              Array [
                Object {
                  "elevation": 4,
                  "height": 150,
                  "padding": 8,
                  "width": 100,
                },
                undefined,
              ]
            }
            keyboardShouldPersistTaps="never"
            onDropdownWillHide={[Function]}
            onDropdownWillShow={[Function]}
            onSelect={[Function]}
            options={
              Array [
                "01",
                "02",
                "03",
                "04",
                "05",
                "06",
                "07",
                "08",
                "09",
                "10",
                "11",
                "12",
                "13",
                "14",
                "15",
                "16",
                "17",
                "18",
                "19",
                "20",
                "21",
                "22",
                "23",
                "24",
                "25",
                "26",
                "27",
                "28",
                "29",
                "30",
                "31",
              ]
            }
            showsVerticalScrollIndicator={true}
            style={
              Array [
                Object {
                  "backgroundColor": "white",
                  "borderColor": "transparent",
                  "borderRadius": 25,
                  "borderWidth": 1,
                  "elevation": 4,
                  "justifyContent": "center",
                },
                undefined,
                Object {},
              ]
            }
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "height": 30,
                      "justifyContent": "center",
                    },
                    Object {
                      "width": 50,
                    },
                  ]
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  disabled={false}
                  ellipsizeMode="tail"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  DD
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#9B9B9B",
            "fontSize": 20,
            "marginHorizontal": 4,
          }
        }
      >
        /
      </Text>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <View
            animated={true}
            defaultIndex={-1}
            defaultValue="Please select..."
            disabled={false}
            dropdownStyle={
              Array [
                Object {
                  "elevation": 4,
                  "height": 150,
                  "padding": 8,
                  "width": 100,
                },
                undefined,
              ]
            }
            keyboardShouldPersistTaps="never"
            onDropdownWillHide={[Function]}
            onDropdownWillShow={[Function]}
            onSelect={[Function]}
            options={
              Array [
                "Jan",
                "Feb",
                "Mar",
                "Apr",
                "Mei",
                "Jun",
                "Jul",
                "Ags",
                "Sep",
                "Okt",
                "Nov",
                "Des",
              ]
            }
            showsVerticalScrollIndicator={true}
            style={
              Array [
                Object {
                  "backgroundColor": "white",
                  "borderColor": "transparent",
                  "borderRadius": 25,
                  "borderWidth": 1,
                  "elevation": 4,
                  "justifyContent": "center",
                },
                undefined,
                Object {},
              ]
            }
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "height": 30,
                      "justifyContent": "center",
                    },
                    Object {
                      "width": 50,
                    },
                  ]
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  disabled={false}
                  ellipsizeMode="tail"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  MM
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
      <Text
        accessible={true}
        allowFontScaling={true}
        disabled={false}
        ellipsizeMode="tail"
        style={
          Object {
            "color": "#9B9B9B",
            "fontSize": 20,
            "marginHorizontal": 4,
          }
        }
      >
        /
      </Text>
      <View>
        <View
          style={
            Object {
              "alignItems": "center",
              "flexDirection": "row",
            }
          }
        >
          <View
            animated={true}
            defaultIndex={-1}
            defaultValue="Please select..."
            disabled={false}
            dropdownStyle={
              Array [
                Object {
                  "elevation": 4,
                  "height": 150,
                  "padding": 8,
                  "width": 100,
                },
                undefined,
              ]
            }
            keyboardShouldPersistTaps="never"
            onDropdownWillHide={[Function]}
            onDropdownWillShow={[Function]}
            onSelect={[Function]}
            options={
              Array [
                "1940",
                "1941",
                "1942",
                "1943",
                "1944",
                "1945",
                "1946",
                "1947",
                "1948",
                "1949",
                "1950",
                "1951",
                "1952",
                "1953",
                "1954",
                "1955",
                "1956",
                "1957",
                "1958",
                "1959",
                "1960",
                "1961",
                "1962",
                "1963",
                "1964",
                "1965",
                "1966",
                "1967",
                "1968",
                "1969",
                "1970",
                "1971",
                "1972",
                "1973",
                "1974",
                "1975",
                "1976",
                "1977",
                "1978",
                "1979",
                "1980",
                "1981",
                "1982",
                "1983",
                "1984",
                "1985",
                "1986",
                "1987",
                "1988",
                "1989",
                "1990",
                "1991",
                "1992",
                "1993",
                "1994",
                "1995",
                "1996",
                "1997",
                "1998",
                "1999",
                "2000",
                "2001",
                "2002",
                "2003",
                "2004",
                "2005",
                "2006",
                "2007",
                "2008",
                "2009",
                "2010",
                "2011",
                "2012",
                "2013",
                "2014",
                "2015",
                "2016",
                "2017",
              ]
            }
            showsVerticalScrollIndicator={true}
            style={
              Array [
                Object {
                  "backgroundColor": "white",
                  "borderColor": "transparent",
                  "borderRadius": 25,
                  "borderWidth": 1,
                  "elevation": 4,
                  "justifyContent": "center",
                },
                undefined,
                Object {},
              ]
            }
          >
            <View
              accessibilityComponentType={undefined}
              accessibilityLabel={undefined}
              accessibilityTraits={undefined}
              accessible={true}
              collapsable={undefined}
              hitSlop={undefined}
              isTVSelectable={true}
              nativeID={undefined}
              onLayout={undefined}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                Object {
                  "opacity": 1,
                }
              }
              testID={undefined}
              tvParallaxProperties={undefined}
            >
              <View
                style={
                  Array [
                    Object {
                      "alignItems": "center",
                      "flexDirection": "row",
                      "height": 30,
                      "justifyContent": "center",
                    },
                    Object {
                      "width": 70,
                    },
                  ]
                }
              >
                <Text
                  accessible={true}
                  allowFontScaling={true}
                  disabled={false}
                  ellipsizeMode="tail"
                  style={
                    Object {
                      "textAlign": "center",
                    }
                  }
                >
                  YYYY
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
